### PR TITLE
Extract rules into separate Markdown files

### PR DESCRIPTION
**Pull request by julianbartel**
Extract rules into separate Markdown files to support individual Frontmatter per rule.
- [x] PowerShell script for building static HTML (PDF) must be extended
- [x] Take rule prefixes from variable in cheat sheet

### CODE REVIEW

File: _layouts/rule-category.html [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+layout: single
+---
+{% assign filtered-rules = site.rules | where_exp: "item", "item.id contains page.category-id" %}
```
Comment (julianbartel) [inline _layouts/rule-category.html]: - The categories are defined implicitly through rule identifiers. Another approach would be to explicitly define rule categories.
 - I could not find out, how to filter rules with a 'starts with' predicate but only with a 'contains' expression

File: _layouts/rule-category.html [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+{% for rule in filtered-rules %}
+<h3 id="{{ rule.rule-id }}">{{ rule.title }} ({{ rule.rule-id }}) <img src="/assets/images/{{ rule.severity }}.png" /></h3>
```
Comment (julianbartel) [inline _layouts/rule-category.html]: - Originally, the Markdown headlines automatically added an id attribute containing the escaped headline text. I changed this so it contains only the ID without text.

**Comment by julianbartel** [comment]
Marked as draft because I would like to clarify a few things (see file comments).
One general point: I thought about removing the AV prefix in rule identifiers, so it could be tailored in forks without need to change every file.

**Comment by dennisdoomen** [comment]
> I thought about removing the AV prefix in rule identifiers, so it could be tailored in forks without need to change every file.

I've used that prefix for the last 15 years, so I'd like to keep it. But maybe we make it a variable?

**Comment by julianbartel** [comment]
> I've used that prefix for the last 15 years, so I'd like to keep it. But maybe we make it a variable?

That's how I meant it. ;-)
I would have done it by simply remove the prefix from the IDs and put it into the template. But a variable is more precise. Just need some time to work that out as Jekyll is new to me.

File: _layouts/rule-category.html (hunk 1 above)
Comment (dennisdoomen) [reply _layouts/rule-category.html]: > * I could not find out, how to filter rules with a 'starts with' predicate but only with a 'contains' expression

I don't think this exists, but you can first _slice_ the first two characters and compare them directly. See https://shopify.github.io/liquid/filters/slice/

> * The categories are defined implicitly through rule identifiers. Another approach would be to explicitly define rule categories.

I think that would be the most clean approach.

**Comment by dennisdoomen** [comment]
> Just need some time to work that out as Jekyll is new to me.

I think you can put that variable, e.g. `rulePrefix`, in the `_config.yml` and then access it through `{{ site.rulePrefix }}`.

**Comment by julianbartel** [comment]
> Search is no longer working

Added rules to the search data store.
> The recent look-and-feel changes somehow are not available in your PR.

Rebased PR on master.

**Comment by dennisdoomen** [comment]
Wow. This really starts to look great. I found another bug though. I'm not sure it's my local environment, bit if you search for `av12`, you can't open any of the results.

**Comment by julianbartel** [comment]
> Wow. This really starts to look great. I found another bug though. I'm not sure it's my local environment, bit if you search for `av12`, you can't open any of the results.

Can you specify this a bit? I made a clean checkout and rebuild to check it out, but search seems to work as expected in my workspace (see screen cap). 🤔

search-sample

**Comment by dennisdoomen** [comment]
@julianbartel awesome work man. It does work as expected. It must have been my machine.

**Comment by dennisdoomen** [comment]
Hmm, just discovered that the "Edit page" button at the top of each page no longer works.

**Pull request merged**